_posts/2021-09-16-drain-cleaner.md

```diff
@@ -0,0 +1,195 @@
+---
+layout: post
+title: "DrainCleaner: what's this?"
+date: 2021-09-16
+author: federico_valeri
+---
+
+One of the promises of Kubernetes is zero downtime upgrades. If your service is highly available, meaning it has more 
+than one replica running on different nodes or availability zones, then it should be possible to do a rolling update 
+without any service disruption.
+
+An event that may affect your service availability is the Kubernetes upgrade, where all cluster nodes are drained one by 
+one, until they are all running on a new version. Node draining is also used for maintenance. When a node is drained, 
+it is cordoned first to prevent scheduling of new Pods and then the old Pods are evicted. This process works well with 
+stateless applications, but can cause problems with stateful and replicated applications like Kafka.
+
+<!--more-->
+
+## The node draining problem
+
+A common use case is a producer application sending records to a topic with 3 replicas and with `min.insync.replica=2`,
+which is required to get resiliency against a single node failure. We want such producer to be able to continue 
+uninterrupted during a Kubernetes upgrade, where each node is drained right after the previous one has been fully 
+evacuated.
+
+The problem is that node draining does not account for the actual internal state of the Kafka cluster, as it works using
+the Readiness state, which indicates if the container is ready to start accepting traffic, but does not necessarily
+correspond to the replicas being in-sync. For that reason, there is no guarantee that the next node draining will start
+before the previous evacuated Kafka Pod is fully synced, so you might end up with under-replicated partitions for some
+time, causing the producer's rate drop to zero. This can affects the entire data pipeline, so also consumers will likely
+be affected.
+
+## Leveraging admission webhooks
+
+The [DrainCleaner](https://github.com/strimzi/drain-cleaner) is a lightweight application that leverage admission 
+webhooks to help moving Kafka Pods from Kubernetes nodes which are being drained. This is useful if you want the 
+ClusterOperator to move Pods instead of Kubernetes itself. The advantage of this approach is that the ClusterOperator 
+makes sure that no partition becomes under-replicated during node draining, by waiting for the synchronization of topics 
+with multiple replicas and `min.insync.replicas < replicas`.
+
+As you can see from the following image, when a new Kubernetes request arrives at the API server endpoint, a 
+[ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) 
+controller is invoked after the schema validation phase and it calls every registered webhook, which implement some 
+logic. This can be useful to enforce custom policies or simply be notified about specific events.
```

Comment:
```suggestion
controller is invoked after the schema validation phase and it calls every registered webhook to implement some 
logic. This can be useful to enforce custom policies or simply be notified about specific events.
```